Resources for Custom TLS and Platform TLS products

bengesoff opened the pull request:
4 new resources:
- `fastly_tls_private_key`
- `fastly_tls_certificate`
- `fastly_tls_platform_certificate`
- `fastly_tls_activation`

10 new data sources
- `fastly_tls_activation`
- `fastly_tls_activation_ids`
- `fastly_tls_certificate`
- `fastly_tls_certificate_ids`
- `fastly_tls_configuration`
- `fastly_tls_configuration_ids`
- `fastly_tls_platform_certificate`
- `fastly_tls_platform_certificate_ids`
- `fastly_tls_private_key`
- `fastly_tls_private_key_ids`

This PR depends on the related `go-fastly` SDK PR (https://github.com/fastly/go-fastly/pull/259) and will need its `go.mod` updating once it gets merged. (Currently uses a `replace` directive to point to the OC fork).

The documentation has been updated to use the new format, thanks to the upstream PRs https://github.com/fastly/terraform-provider-fastly/pull/356 and https://github.com/fastly/terraform-provider-fastly/pull/362. This carries with it the same caveats as noted elsewhere with regards to the generated documentation for nested schemas.

There were a few tweaks added on top of the docs generation script. In particular I added some logic to the Makefile to install `tfplugindocs` to a project-local `./bin` directory due to the volatility in the generated documentation. I also removed the module caching from the github workflow because the tool was installed from the `vendor` directory which was already inside the git repo.

The Makefile had a command added for [sweepers](https://www.terraform.io/docs/extend/testing/acceptance-tests/sweepers.html) as well.

wjam committed:
TLS Configuration Data Source

bengesoff committed:
TLS Custom Private Key Resource

wjam committed:
TLS Custom Certificate Resource

bengesoff committed:
TLS Custom Private Key Data Source

wjam committed:
TLS Custom Certificate Data Source

Also add sweepers for TLS certificates and private keys to easily clean up resources leaked during any failed tests.

bengesoff committed:
TLS Activation Resource

bengesoff committed:
TLS Activation Data Source

wjam committed:
TLS Private Key IDs Data Source

bengesoff committed:
TLS Custom Certificate IDs Data Source

bengesoff committed:
TLS Configuration IDs Data Source

Commit:
TLS Activation IDs Data Source

bengesoff committed:
Change test names to use underscore

Terraform testing style guide seems to suggest camel case is used for
the main test name then an underscore separates different variations of
it.

wjam committed:
TLS Platform Certificate Resource

bengesoff committed:
Update go-fastly reference based on merged PR

bengesoff committed:
TLS Platform Certificate Data Source

bengesoff committed:
TLS Platform Certificate IDs Data Source

bengesoff committed:
Platform TLS Data Sources documentation

bengesoff committed:
Remove redundant `id` filter function

bengesoff committed:
Add dns_records attribute to TLS Configuration Data Source

bengesoff committed:
Bring in changes from upstream and modify new code to match

Main changes were moving docs generation to tfplugindocs, and updating
the go-fastly SDK to v3. I added some changes to the upstream docs
generation to avoid having to globally install tfplugindocs. This was
also done upstream so I had to do some large merge conflict resolution
in this commit to combine the similar but different updates.

One commit message related to vendoring tfplugindocs was:

> Don't cache dependencies in github PR workflow, instead rely on /vendor
>
> Including the tfplugindocs module in vendor means it's updating with `go
> mod vendor` along with the other libraries used. When running `go
> install`, this vendored copy is used, and installed to a project-local
> /bin directory. This enables the version of tfplugindocs used to be
> independent of other go projects installed on one's system.
>
> This change means `make dependencies` is no longer used, and isn't
> needed in the github PR workflow. Furthermore, the source code for the
> tool is included in the /vendor already so the caching of ~/go/* isn't
> required either.

kellymclaughlin committed:
Add TLSCLientCert and TLSClientKey options for splunk logging (#353)

* Add TLSCLientCert and TLSClientKey options for splunk logging

* Add some comments to clarify the usage splunk test tls cert values

* Update fastly/block_fastly_service_v1_splunk_test.go

* Update fastly/block_fastly_service_v1_splunk_test.go

* Update fastly/block_fastly_service_v1_splunk_test.go

* Update fastly/block_fastly_service_v1_splunk_test.go

* Update fastly/block_fastly_service_v1_splunk_test.go

* Update fastly/block_fastly_service_v1_splunk_test.go

* Update fastly/block_fastly_service_v1_splunk_test.go

Co-authored-by: Mark McDonnell <Integralist@users.noreply.github.com>

bengesoff committed:
Stylistic tweaks to make resources more aligned

A couple naming/structure things resulting from different people writing the code. Have just tidied them up before PRing.

bengesoff committed:
Update go-fastly reference

bengesoff committed:
Merge remote-tracking branch 'upstream/master' into tls-platform-and-custom

bengesoff commented on fastly/validators.go:
```diff
@@ -140,3 +141,16 @@ func validateHTTPSURL() schema.SchemaValidateFunc {
+
+func validatePEMBlock(pemType string) schema.SchemaValidateFunc {
+	return func(val interface{}, key string) ([]string, []error) {
+		b, _ := pem.Decode([]byte(val.(string))) // TODO: might want to loop here to support multiple PEM resources in one string for intermediates blob?
```
Outstanding question here about what the contents of the `intermediates_blob` should be. Currently only validates for one certificate but maybe there could be multiple? I asked Alexander on Friday who said he'd check

bengesoff replied:
I've just had a response - the suspicion was correct in that there can be multiple PEM blocks in the `intermediates_blob`. Will need to update to account for this, either as an extra on this PR or a followup depending on timings

bengesoff commented on go.mod:
```diff
@@ -2,11 +2,12 @@ module github.com/fastly/terraform-provider-fastly
 
 go 1.14
 
+replace github.com/fastly/go-fastly/v3 v3.0.0 => github.com/opencredo/go-fastly/v3 v3.0.0-20210208131624-9ccb395d6808
```
Temporary redirect to OC fork until https://github.com/fastly/go-fastly/pull/259 is merged

bengesoff commented on docs/data-sources/tls_configuration.md:
```diff
@@ -0,0 +1,56 @@
+- **updated_at** (String) Timestamp (GMT) when the configuration was last updated.
+
+<a id="nestedatt--dns_records"></a>
+### Nested Schema for `dns_records`
```
Same issue as @Integralist raised previously (https://github.com/hashicorp/terraform-plugin-docs/issues/28) - no description on nested schema. I've left it autogenerated because I didn't think the descriptions were too critical, it is reasonably self-explanatory what the fields are for, and it seemed easier to maintain by leaving it autogenerated with the rest of the docs.

bengesoff committed:
Support multiple certificates in platform certificate intermediates_blob

The intermediates_blob field of the fastly_tls_platform_certificate
resource can contain PEM blocks representing an arbitrary length chain
of certificates. The validation function for this field has been updated
to reflect this. It now loops through the provided string and checks
that each block it finds matches the expected block type until it
reaches the end of the string. Similarly the validation function for one
single block has been updated to fail if the string contains more than
one PEM block.

bengesoff committed:
Use allow_untrusted_root in platform certificate update function

Was only used in creation function but should have also been used
in update too.

Integralist commented on scripts/generate-docs.go:
```diff
@@ -144,7 +200,11 @@ func getTemplate(tmplDir string) *template.Template {
 		return nil
 	})
-	return template.Must(template.ParseFiles(templateFiles...))
+	template, err := template.ParseFiles(templateFiles...)
```
Ah I didn't notice `template.ParseFiles` returned an error. Good spot in that I wouldn't have been catching any error and so the code would have imploded before reaching `template.Must`.

bengesoff replied:
Yes I think I was having some trouble with it when I was moving some of the old style templates over and it wasn't happy about something in the parsing step so it helped to debug that - figured it'd probably be helpful for someone else too.

Integralist commented on scripts/generate-docs.go:
```diff
@@ -74,6 +74,46 @@ func main() {
 			name: "ip_ranges",
 			path: tempDir + "/data-sources/ip_ranges.md.tmpl",
 		},
+		{
```
It would be nice now that I think about it for these slices to be dynamically populated rather than having to remember to update the script every time a new resource is added.

Integralist replied:
But I wouldn't worry too much about it for this PR.

bengesoff replied:
Yeah that could be a nice enhancement - scanning the directories or something

Integralist commented on .github/workflows/pr.yml:
```diff
@@ -52,25 +52,6 @@ jobs:
       uses: actions/setup-go@v2
       with:
         go-version: 1.14.x
-    - name: Restore cached binaries
-      id: cache
-      uses: actions/cache@v2
-      with:
-        path: ~/go/bin
-        key: ${{ runner.os }}-go-bin-${{ hashFiles('**/go.sum') }}
-        restore-keys: |
-          ${{ runner.os }}-go-bin-
-    - name: Restore cached modules dependencies
-      uses: actions/cache@v2
-      with:
-        path: ~/go/pkg/mod
-        key: ${{ runner.os }}-go-mod-${{ hashFiles('**/go.sum') }}
-        restore-keys: |
-          ${{ runner.os }}-go-mod-
-    - name: Install dependencies
-      if: steps.cache.outputs.cache-hit != 'true'
-      run: make dependencies
-      shell: bash
```
Any explanation you can give me for why these segments of the workflow have been deleted?

bengesoff replied:
Yes this change tied in with some of the other changes relating to the way `tfplugindocs` was being installed. I think previously the makefile was installing the package to `GOROOT` (i.e. `~/go/pkg` and `~/go/bin`), which would explain the steps in the workflow caching these directories. 

However when adapting the upstream docs generation changes I found that the output of the `tfplugindocs` tool was quite volatile, given its alpha status, and thought it made sense to use the `vendor` style dependency management to pin in the exact version of the source code used to generate the docs and reduce disparity between all of the contributor's versions of it. In a similar vein I changed it to install the binary to a project-local `./bin` directory to isolate it in case other terraform plugin projects were using different versions of it on anyone's machine. (I think you also made some very similar changes in parallel, e.g. adding a `tools.go` to allow `go mod` to find it, so I ended up merging my tweaks into that as well).

Anyway long story short, with the source code being installed to `./vendor` and the compiled `tfplugindocs` binary being installed to `./bin`, I didn't see a need to manually cache the external global `GOROOT` directories as the GitHub workflow does. (Also it had stopped working for me when I removed the `make dependencies` target, so I had to fix it somehow! 😉)

Does this make sense do you think? Happy to jump on a call and discuss it further if you think that would be helpful

Integralist replied:
No that's fine. Thanks for the background information 👍🏻

Integralist commented on .gitignore:
```diff
@@ -7,6 +7,7 @@ terraform.tfstate
 bin/
 modules-dev/
 /pkg/
+playground/
```
What does this folder relate to?

bengesoff replied:
Not sure I think someone must have been experimenting - I'll remove it

bengesoff committed:
Update go-fastly reference to v3.3.0

Also removed the `replace` directive in the go.mod to remove dependency
on opencredo fork.

Integralist commented on GNUmakefile:
```diff
@@ -52,18 +52,20 @@ test-compile:
-
-install-tools: dependencies
+BIN=$(CURDIR)/bin
+$(BIN)/%:
```
Does this need to be a dynamic target? Could the target still be named `install-tools` and `generate-docs` to reference it? My thinking is that in the updated `generate-docs` we're still adding `$BIN` to `$PATH` so the `go install`'ed tools will still be located there. 

I feel like the dynamic target name (and specifically setting `$(BIN)/tfplugindocs`) is too much generalization without any real value? Although I could be mis-understanding the purpose for this refactor entirely.

bengesoff replied:
This was related to the same change as explained in the comment on `.github/workflows/pr.yml`. As part of moving the installation of `tfplugindocs` to be inside the `terraform-provider-fastly` directory, I used the non-PHONY target to avoid needing to re-run it every time, if the binary already exists. By extension, the dynamic target seemed useful in case other dependencies were added, given that the grep/awk command was already capable of installing the whole `tools.go` file.

That's the rationale anyway - I'm happy to change it to use a PHONY `install-tools` target if you'd prefer, and if I'm not mistaken that would rebuild the tool on each invocation, which could be helpful if it were upgraded? That's assuming that you agree installing to a project-local `terraform-provider-fastly/bin` directory is a good idea anyway, which you might not - also ok!

Integralist replied:
OK that makes sense. I think avoiding reinstalling the tool every time is better any way and we can suffer not being on the bleeding edge all the time.

Integralist commented on GNUmakefile:
```diff
@@ -52,18 +52,20 @@ test-compile:
-generate-docs: install-tools
-	go run scripts/generate-docs.go
+validate-docs: $(BIN)/tfplugindocs
+	$(BIN)/tfplugindocs validate
```
Would this not also need `PATH=$(PATH):$(BIN)` ? As you might not necessarily run `generate-docs` first.

bengesoff replied:
I don't believe so, as it's calling the binary from its relative filepath instead of letting the shell look it up on the system `PATH`, but I might be wrong. In the case of `generate-docs` the script calls it without specifiying the location so the `PATH` is needed there. 

Now you mention it, it could be cleaner to specify `./bin/tfplugindocs` inside the script to avoid modifying the `PATH` for that command, if we decide that a non-global install is the right way to go. Or if could be nice to leave it configurable from the Makefile too so the location is saved in one place, in which case maybe passing an argument in could be best.

bengesoff replied:
Just a heads up here - I've pushed in a change to avoid modifying `PATH` for the `generate-docs`

Integralist commented on docs/data-sources/tls_activation.md:
```diff
@@ -0,0 +1,38 @@
+
+### Read-Only
+
+- **created_at** (String) Timestamp (GMT) when TLS was enabled.
```
Interesting this is GMT? I would have thought UTC would have been the preferred timezone 🤔 
Is the API determining this or is this the provider code that's configured to GMT?

Integralist replied:
Same comment effectively everywhere we use GMT.

bengesoff replied:
This was taken directly from the API docs I believe!

Integralist replied:
Fair enough then 🙂

Integralist commented on fastly/data_source_fastly_tls_configuration_ids.go:
```diff
@@ -0,0 +1,44 @@
+	// 2.x upgrade note - `hashcode.String` was removed from the SDK
+	// Code will need to be copied into this repository
+	// https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html#removal-of-helper-hashcode-package
+	d.SetId(fmt.Sprintf("%d", hashcode.String(""))) // hashCode should include any filters set when they are added
```
What's the rationale behind passing an empty string to `hashcode.String`?

bengesoff replied:
If I recall correctly, this was mainly as a placeholder anticipating the addition of filters to this data source at some point, similarly to how the [`aws_ami_ids`](https://github.com/hashicorp/terraform-provider-aws/blob/bc48c2178fff1c5b1b142481069e45fcb5870bd5/aws/data_source_aws_ami_ids.go#L111) data source does it, but in this case there are no filters so an empty string was used in the mean time. It could be hardcoded to `0` if you think that would be cleaner, and a `hashcode` or similar could be added in at a later point if additional filters are added?

Integralist replied:
> anticipating the addition of filters to this data source at some point

if it's likely we're going to see filters come in the future, then I think this is fine to leave 'as is', otherwise I'd suggest we set some kind of value to be hashed.

Integralist commented on docs/data-sources/tls_private_key_ids.md:
```diff
@@ -0,0 +1,31 @@
+
+### Optional
+
+- **id** (String) The ID of this resource.
```
Probably a silly question, but I've seen in a few places this pattern of having an optional `id` attribute for a data source that simply returns multiple data records, and I'm a bit confused by the purpose of it. e.g. when would you want to pass an id for this data source?

bengesoff replied:
Good spot, that seems to be an artifact of `tfplugindocs` - I agree there isn't a purpose of actually specifying this. What do you think the policy generally should be on tfplugindocsisms like this?

Integralist replied:
Hmm, I guess we just leave it in if that's what the plugin is doing (I mean it is the _official_ tool now so it makes sense not to fight against it).

Integralist commented on docs/resources/tls_activation.md:
```diff
@@ -0,0 +1,75 @@
+
+resource "fastly_tls_activation" "test" {
+  certificate_id = fastly_tls_certificate.demo.id
+  domain = "example.com"
```
Just out of interest, but would it possible to reference the `example.com` already defined, via `fastly_service_v1.demo.domain.name` or maybe `fastly_service_v1.demo.domain[0].name` ?

bengesoff replied:
I'm not sure this would work, because `fastly_service_v1.demo.domain` refers to a set, not a list, and so can't be indexed directly. Alternatively `tolist(fastly_service_v1.test.domain)[0].name` could be used, or `[for d in fastly_service_v1.domain : d. name][0]` etc. Whichever you think is clearest between these or hardcoding 🙂

Integralist replied:
OK good to know. I asked only for my own learning of the terraform hcl syntax (of which I'm still very green on). 👍🏻

Integralist commented on templates/resources/tls_certificate.md.tmpl:
```diff
@@ -0,0 +1,63 @@
+
+Uploads a custom TLS certificate to Fastly to be used to terminate TLS traffic.
+
+-> Each TLS certificate **must** have its corresponding private uploaded _prior_ to uploading the certificate. This can
```
```suggestion
-> Each TLS certificate **must** have its corresponding private key uploaded _prior_ to uploading the certificate. This can
```

Integralist commented on templates/data-sources/tls_activation_ids.md.tmpl:
````diff
@@ -0,0 +1,31 @@
+## Example Usage
+
+```hcl
+
````
Can we remove the extraneous empty lines at the top/bottom of this code snippet.

Integralist commented on templates/resources/tls_platform_certificate.md.tmpl:
```diff
@@ -0,0 +1,98 @@
+
+Uploads a TLS certificate to the Fastly Platform TLS service.
+
+-> Each TLS certificate **must** have its corresponding private uploaded _prior_ to uploading the certificate. This can
```
```suggestion
-> Each TLS certificate **must** have its corresponding private key uploaded _prior_ to uploading the certificate. This can
```

Integralist commented on fastly/data_source_fastly_tls_activation.go:
```diff
@@ -0,0 +1,167 @@
+				Optional:      true,
+				Computed:      true,
+				Description:   "Fastly Activation ID. Conflicts with all other filters.",
+				ConflictsWith: []string{"certificate_id", "configuration_id", "domain"},
```
I'm unclear on what the purpose of this field is.

> ConflictsWith is a set of schema keys that conflict with this schema. This will only check that they're set in the _config_. This will not raise an error for a malfunctioning resource that sets a conflicting key.

So sounds like the field doesn't do any validation other than ensuring they're _all_ set (from I can tell).

Put another way, what would happen if this line was removed? (as we'd still be checking for the conflicting keys in `getTLSActivationFilters()` if `id` was not provided).

That said, although we're using `d.GetOk` to check for the conflicting keys (which we're relying on in the absence of `id`), is the point of `ConflictsWith` to have terraform error if the user doesn't provide at least one of these fields? But then would we not prefer to use `AtLeastOneOf` rather than expecting them to provide _all_ three alternative attributes?

bengesoff replied:
We used this pattern in all the singular data sources, I believe, to enforce that _either_ the `id` field is used, _or_ 0-3 of the other filters, but not both. This is so that the `id` field can be used in conjunction with the corresponding plural data source, in which case it wouldn't make any sense to specify another filter as it's just a simple ID lookup. Similarly, it wouldn't make sense for a user to specify an ID if they were trying to filter based on another attribute.

So you're right in that it's just to get a nice Terraform error for the user, but it's less about enforcing they're all set, just that none of those listed are also set when `id` is present. Put differently, `ConflictsWith` doesn't care how many of the listed fields are set, as long as none of them are set when `id` is.

Integralist replied:
> ConflictsWith doesn't care how many of the listed fields are set, as long as none of them are set when id is.

Gotcha. Thanks! 👍🏻

Integralist commented on fastly/data_source_fastly_tls_activation.go:
```diff
@@ -0,0 +1,167 @@
+}
+
+func dataSourceFastlyTLSActivationRead(d *schema.ResourceData, meta interface{}) error {
+
```
See my comment below about appropriate line spacing (and presume I've made the same suggestion elsewhere, especially as it seems we're being inconsistent with the spacing)...

```suggestion
```

bengesoff replied:
Good spot, thanks!

Integralist commented on fastly/data_source_fastly_tls_activation.go:
```diff
@@ -0,0 +1,167 @@
+	var activation *fastly.TLSActivation
+
+	if v, ok := d.GetOk("id"); ok {
+
```
I'm ok with breathing space around blocks of code, but I'm not sure I'm a fan of extra lines like is shown here where we have a line space after the if conditional and again before the else block. I'd prefer to see those sorts of line spacing removed.

bengesoff replied:
Agreed, I'm afraid I didn't notice these, will update.

Integralist commented on fastly/data_source_fastly_tls_activation.go:
```diff
@@ -0,0 +1,167 @@
+}
+
+func listTLSActivations(conn *fastly.Client, filters ...TLSActivationPredicate) ([]*fastly.TLSActivation, error) {
+	var activations []*fastly.TLSActivation
```
Is there going to be a future concern of holding a potentially unbounded number of results in-memory?

Considering we appear to only care about the zero-index from the collection, should we not break the loop after we've received a single activation result?

bengesoff replied:
This is a good point, yes. In the case of TLS activations we do not share the `list*` function between the singular and plural data sources (because the plural one uses an API filter directly whereas the others do all the filtering after pulling the results down). Therefore in this case, we could just modify this function to break the loop early once the filter has passed. However, in most (if not all) of the other data sources, we do share the `list*` function between singular and plural data sources, to abstract away the API's inconsistent pagination methods, so we might lose the consistency among the data sources in this case.

Do you think the consistency with other data sources takes precedence, or should we take the easy code optimisation here while we can?

Integralist replied:
I think consistency is fine in this case. If there's a performance issue that arises we can address it when it happens.

Integralist commented on fastly/data_source_fastly_tls_activation_ids.go:
```diff
@@ -0,0 +1,73 @@
+				Type:        schema.TypeSet,
+				Computed:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Set:         schema.HashString,
```
Is there a reason we've used `Set` here? I've noticed we've not defined `Set` on other uses of schema.TypeSet in our provider code, so I'm wondering if that's a mistake on our part of if there's a reason why you've used it here.

My understanding is that this function (`schema.HashString`) is used internally by terraform to ensure (i.e. validate) our set data structure definitely consists of unique values.

bengesoff replied:
Well spotted - I've just looked this up and it's apparently not needed anymore as Terraform:

> now does the right thing, even with nested data structures

so I will go ahead and remove these!

Integralist commented on fastly/data_source_fastly_tls_activation_ids_test.go:
```diff
@@ -0,0 +1,107 @@
+)
+
+func TestAccDataSourceFastlyTLSActivationIds_basic(t *testing.T) {
+	domain := fmt.Sprintf("tf-test-%s.com", acctest.RandomWithPrefix("tf-test-"))
```
If I'm understanding the `acctest.RandomWithPrefix` function, I presume what we actually want is...

```suggestion
	domain := fmt.Sprintf("%s.com", acctest.RandomWithPrefix("tf-test-"))
```

...as the random name we're generating will already have `"tf-test-"` as part of the prefix.

Integralist replied:
Same comment applies in other test files.

Integralist commented on fastly/sweeper_test.go:
```diff
@@ -0,0 +1,48 @@
+
+var sweeperClients map[string]*fastly.Client
+
+func TestMain(m *testing.M) {
```
I read the docs on `resource.TestMain` and I still don't understand how it works or what it tests 🤦🏻

Integralist replied:
I see you assign to `sweeperClients` but then in the other functions (called from other tests) it relies on `sweeperClients` being defined so how do we ensure `TestMain` is run ahead of those other tests that also call functions like `sharedClientForRegion`?

bengesoff replied:
I'll be honest I haven't really looked into how it works myself - I saw it looked very similar to [what the docs recommend](https://www.terraform.io/docs/extend/testing/acceptance-tests/sweepers.html) and left it at that! I believe the testing framework (called by the Makefile with `go test ./fastly -v -sweep=ALL ........`) takes care of running `TestMain` etc in the correct order.

If you're asking more about what the purpose is, the link above explains it quite well, but essentially it's just a "cleanup" script to delete any leaked resources that were provisioned by a test and that didn't get correctly deleted. You can run `make sweep` at some point after running the tests and it'll take care of deleting anything that matches a certain criteria, which in this case is the `testResourcePrefix` prefix

Integralist commented on fastly/data_source_fastly_tls_certificate.go:
```diff
@@ -0,0 +1,223 @@
+
+type TLSCertificatePredicate func(*fastly.CustomTLSCertificate) bool
+
+func getTLSCertificateFilters(d *schema.ResourceData) []TLSCertificatePredicate {
```
There's a lot of similarities in this file to the tls_activation file. I'm guessing without having generics in the language there's no _reasonable_ away to consolidate the duplication?

bengesoff replied:
This was definitely one that seemed a good candidate for generics, agreed! I don't believe so, no, if we maintain the "_reasonable_" caveat. Off the top of my head the main option would be to pass around `interface{}`s instead of the `fastly.*` types and either forfeit type safety or have a load of type switches around obscuring the logic. If you think of anything, let me know, otherwise I guess we're waiting until December or whenever generics gets added!

Integralist replied:
👍🏻

Integralist commented on fastly/data_source_fastly_tls_certificate_ids.go:
```diff
@@ -0,0 +1,46 @@
+	// 2.x upgrade note - `hashcode.String` was removed from the SDK
+	// Code will need to be copied into this repository
+	// https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html#removal-of-helper-hashcode-package
+	d.SetId(fmt.Sprintf("%d", hashcode.String(""))) // hashCode should include any filters set when they are added
```
I don't think I understood the comment:

> // hashCode should include any filters set when they are added

bengesoff replied:
Maybe the wording isn't clear - it's supposed to indicate that, if additional filters are added to the data source (such as `tls_activation_ids`) then the filters should be used instead of the empty string. I'll try and clear this up

Integralist replied:
Awesome, thanks!

Integralist commented on fastly/data_source_fastly_tls_configuration.go:
```diff
@@ -0,0 +1,292 @@
+	}
+
+	d.SetId(configuration.ID)
+	if err := d.Set("name", configuration.Name); err != nil {
```
When would we want to use partial state? e.g. `d.SetPartial(...)`

I've typically seen it used when dealing with 'resources', but does it ever make sense to use it with 'data sources'? and isn't that sort of what's happening here anyway (e.g. we're only getting partial data back and set into the local state before returning an error)? or does the data not get set into the state file if there is an error returned?

bengesoff replied:
I don't think it's ever needed as it's deprecated in SDK 1.x and removed in SDK 2.x. If there's an error then the state should get refreshed on the next run anyway so there shouldn't be any need to partially fill the state if I'm not mistaken? The only edge case I can possibly think of is if there is some one-time piece of information that won't be available on the next call to `Read` but I don't believe that applies anywhere here.

Integralist replied:
> removed in SDK 2.x

oh really! lulz

Integralist commented on fastly/data_source_fastly_tls_configuration_ids_test.go:
```diff
@@ -0,0 +1,21 @@
+	})
+}
+
+const testAccFastlyAccDataSourceTLSConfigurationIDs = `data "fastly_tls_configuration_ids" "subject" {}`
```
I typically have constants at the top of the file. Is there a benefit to have it underneath the code that calls it?

I guess it comes down to how you prefer to read code vs what makes reading code easier.

bengesoff replied:
I think this is underneath because it's a test piece of HCL config and they seem to normally be at the bottom of the files when they are functions, so would make sense to group them even if some of them don't have any variable data. Happy to shift it around if you'd prefer.

Integralist replied:
No that's fine. I didn't really pay attention to the fact it was mock hcl, and I agree that those always appear to be at the bottom of the test file 👍🏻

Integralist commented on fastly/data_source_fastly_tls_configuration_ids_test.go:
```diff
@@ -0,0 +1,21 @@
+		Steps: []resource.TestStep{
+			{
+				Config: testAccFastlyAccDataSourceTLSConfigurationIDs,
+				Check:  resource.TestCheckResourceAttrSet("data.fastly_tls_configuration_ids.subject", "ids.#"),
```
I presume this is just validating that there are zero elements within the ids list, is my understanding correct?

bengesoff replied:
I think it's just validating that the `ids` field has an attribute for its length and therefore loaded without error, not that it is specifically zero. There's not really anything interesting to test here because we don't know what configurations, if any, might be present and we can't create one and look for it, as we do in the other resources (eg activations)

Integralist commented on fastly/provider.go:
```diff
@@ -24,17 +24,31 @@ func Provider() terraform.ResourceProvider {
+			"fastly_tls_platform_certificate":     dataSourceFastlyTLSPlatformCertificate(),
+			"fastly_tls_platform_certificate_ids": dataSourceFastlyTLSPlatformCertificateIDs(),
+			"fastly_tls_private_key":              dataSourceFastlyTLSPrivateKey(),
+			"fastly_tls_private_key_ids":          dataSourceTLSPrivateKeyIDs(),
```
Should this not be named consistently with the other data sources?
```suggestion
			"fastly_tls_private_key_ids":          dataSourceFastlyTLSPrivateKeyIDs(),
```

That said, I notice the 'resources' listed below aren't named consistently in that they don't mention the provider name `Fastly`. I'm not suggesting we change that as part of this PR as it's an unrelated refactoring, but more from an idiomatic terraform code perspective shouldn't they be `<resource|dataSource>Fastly<Name>`?

e.g.

```diff
- resourceServiceV1
+ resourceFastlyServiceV1
```

bengesoff replied:
Ahh yes good point, I did this check on filenames before opening the PR but missed the function names 😩

Integralist commented on fastly/data_source_fastly_tls_private_key_ids.go:
```diff
@@ -0,0 +1,47 @@
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func dataSourceTLSPrivateKeyIDs() *schema.Resource {
```
Should this not be named consistently with the other data sources?
```suggestion
func dataSourceFastlyTLSPrivateKeyIDs() *schema.Resource {
```

I presume we'd want to update the READ function as well: 

```diff
- dataSourceTLSPrivateKeyIdsRead
+ dataSourceFastlyTLSPrivateKeyIdsRead
```

Integralist commented on fastly/data_source_fastly_tls_private_key.go:
```diff
@@ -0,0 +1,189 @@
+
+func dataSourceFastlyTLSPrivateKey() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceTLSPrivateKeyRead,
```
Should we keep the CRUD function names consistent with the parent resource?

e.g.
```suggestion
		Read: dataSourceFastlyTLSPrivateKeyRead,
```

Integralist commented on fastly/resource_fastly_tls_activation_test.go:
```diff
@@ -0,0 +1,160 @@
+	return func(state *terraform.State) error {
+		conn := testAccProvider.Meta().(*FastlyClient).conn
+
+		r := state.RootModule().Resources[resourceName]
```
Isn't there a possibility this will cause a runtime panic if there is no `resourceName` key defined? I appreciate the test will fail and that's a good thing, but that maybe we should handle that error scenario a bit more cleanly if so. What do you reckon?

Integralist reviewed (commented):
@bengesoff this PR looks great! I've left some questions and minor suggestions. Let me know your thoughts when you get a chance.

bengesoff committed:
First updates from PR feedback

- removal of unneeded .gitignore entry
- removal of superfluous whitespace in docs example block
- conversion of TypeList to TypeSet in plural data sources' `ids` field
- a couple typo fixes here and there
- removal of Set function for controlling set hashing, unneeded
- consolidation of function naming to include "Fastly" before resource
  name
- fix some acctest.RandomWithPrefix with duplicate prefix
- clarify some comments
- add some checks in testAcc.*Exists functions when accessing the map of
  resources in state to avoid a panic if resource not found

bengesoff commented:
@Integralist I have fixed&resolved some comments, and replied to the rest. Thanks for the feedback, caught some things that had slipped through and raised some good points!

bengesoff committed:
Make tfplugindocs location configurable with flag instead of PATH

Add a -tfplugindocsPath command line argument to the parsing script to
make it a bit more robust than dynamically setting the PATH variable in
the Makefile. Defaults to local bin, as the Makefile expects, but I
still set the argument in the Makefile in case someone modifies the BIN
variable.

Integralist reviewed (approved):
@bengesoff LGTM.

I'll get this merged and then if you could let me know once you've rebased the latest master into your #365 PR.

Thanks!
